flare/ase/otf.py

```diff
@@ -1,358 +1,182 @@
-            self.call_observers()
-            curr_struc = Structure.from_ase_atoms(self.atoms)
-            self.l_bound = get_l_bound(self.l_bound, curr_struc, self.two_d)
-            print('l_bound:', self.l_bound)
-            curr_struc.stds = np.copy(self.stds)
-            noise = calc.gp_model.hyps[-1]
-            self.std_in_bound, self.target_atoms = is_std_in_bound(\
-                    self.std_tolerance_factor, noise, curr_struc, self.max_atoms_added)
-
-            print('std in bound:', self.std_in_bound, self.target_atoms)
-
-            if not self.std_in_bound:
-                # call dft/eam
-                print('calling dft')
-                dft_forces = self.call_DFT()
-
-                # update gp
-                print('updating gp')
-                self.update_GP(dft_forces)
-                calc.mgp_updated = False
-
-            if self.use_mapping:
-                self.build_mgp()
-
-        self.observers[self.logger_ind][0].run_complete()
-
-
-    def build_mgp(self):
-        # build mgp
-        calc = self.atoms.calc
-        if self.nsteps in self.non_mapping_steps:
-            calc.use_mapping = False
-            skip = True
-        else:
-            calc.use_mapping = True
+    def initialize_train(self):
+
+        super().initialize_train()
 
-            if calc.mgp_updated:
-                skip = True
+        if self.md_engine == 'NPT':
+            if not self.md.initialized:
+                self.md.initialize()
             else:
-                skip = False
-                calc.mgp_updated = True
+                if self.md.have_the_atoms_been_changed():
+                    raise NotImplementedError(
+                        "You have modified the atoms since the last timestep.")
+        
 
-        calc.build_mgp(skip)
+    def compute_properties(self):
+        '''
+        compute forces and stds with FLARE_Calculator
+        '''
+        if not isinstance(self.atoms.calc, FLARE_Calculator):
+            self.atoms.set_calculator(self.flare_calc)
 
+        self.atoms.calc.results = {}
+        f = self.atoms.get_forces(self.atoms)
+        stds = self.atoms.get_uncertainties(self.atoms)
+        self.structure.forces = deepcopy(f) 
+        self.structure.stds = deepcopy(stds)
 
-    def call_DFT(self):
-        self.dft_calc.nsteps = self.nsteps
-        prev_calc = self.atoms.calc
-        calc = deepcopy(self.dft_calc)
-        self.atoms.set_calculator(calc)
-        forces = self.atoms.get_forces()
-        self.call_observers()
-        self.atoms.set_calculator(prev_calc)
-        self.dft_count += 1
-        return forces
 
-    def update_GP(self, dft_forces):
-        atom_count = 0
-        atom_list = []
-        gp_model = self.atoms.calc.gp_model
+    def md_step(self): 
+        '''
+        Get new position in molecular dynamics based on the forces predicted by
+        FLARE_Calculator or DFT calculator
+        '''
+        self.md.step()
```

Comment:
Somehow, self.md.step() updates the positions of self.structure, even though it should act only on self.atoms. This causes the velocities and temperatures to be reported as zero in the OTF output file.

Reply:
1829b0c seems to give a temporary fix to this problem by manually resetting the positions of self.structure. But I'm still confused why self.md.step() updates self.structure.positions...

Reply:
The issue was that self.md.step() returned atoms.positions rather than a copy of it, which caused self.atoms and self.structure to get updated together downstream. a2a785a fixes this